ligo/haskell/src/Ligo/Typescript.hs

```diff
@@ -0,0 +1,380 @@
+toEpType (NTInt _) = EpInt
+toEpType (NTNat _) = EpNat
+toEpType (NTUnit _) = EpUnit
+toEpType (NTSignature _) = EpSignature
+toEpType (NTOption _ a) = EpOption (toEpType a)
+toEpType (NTOr _ n1 n2 f1 f2) = EpOr (n1, toEpType f1) (n2, toEpType f2)
+toEpType (NTPair _ n1 n2 f1 f2) = EpPair (n1, toEpType f1) (n2, toEpType f2)
+toEpType (NTKey _) = EpKey
+toEpType (NTChainId _) = EpChainId
+toEpType (NTList _ a) = EpList (toEpType a)
+toEpType (NTSet _ a) = EpSet (toEpType a)
+toEpType (NTOperation _) = EpOperation
+toEpType (NTMap _ f1 f2) = EpMap (toEpType f1) (toEpType f2)
+toEpType (NTBigMap _ f1 f2) = EpBigMap (toEpType f1) (toEpType f2)
+toEpType (NTContract _ a) = EpContract (toEpType a)
+toEpType (NTLambda _ f1 f2) = EpLambda (toEpType f1) (toEpType f2)
+toEpType (NTBytes _) = EpBytes
+toEpType (NTMutez _) = EpMutez
+toEpType (NTKeyHash _) = EpKeyHash
+toEpType (NTTimestamp _) = EpTimestamp
+toEpType (NTAddress _) = EpAddress
+toEpType (NTBool _) = EpBool
+
+-- looks up the entrypoints in the Notes t for the whole parameter.
+-- It might be possible to do this without re-implementing logic in
+-- the `mkEntrypointsMap` from morley.
+lookupEntryPointsAndTypes :: Notes t -> [Entrypoint]
+lookupEntryPointsAndTypes (NTOr _ fa1 fa2 f1 f2) =
+  let
+    leftEntrypoint = case extractAnn fa1 of
+      Just ann_ -> (Entrypoint ann_ (toEpType f1)) : (lookupEntryPointsAndTypes f1)
+      Nothing -> lookupEntryPointsAndTypes f1
+    rightEntrypoint = case extractAnn fa2 of
+      Just ann_ -> (Entrypoint ann_ (toEpType f2)) : (lookupEntryPointsAndTypes f2)
+      Nothing -> lookupEntryPointsAndTypes f2
+  in leftEntrypoint <> rightEntrypoint
+lookupEntryPointsAndTypes _ = []
+
+-- Check if the annotation in this annotation is among the
+-- ones we are not interested in, or empty. If so return a Nothing. Else
+-- wrap the annotation in a Just before returning it.
+extractAnn :: Annotation tag -> Maybe Text
+extractAnn ann_ =
+  case (ann_ == noAnn) || (isJust $ find (== ann_) ignoreEntypoints) of
+    True -> Nothing
+    _ -> Just $ unAnnotation ann_
+
+-- Some entrypoints we don't want to call directly.
+ignoreEntypoints :: [Annotation tag]
+ignoreEntypoints = [ann "running", ann "transfer_contract_tokens", ann "xtzForbidden", ann "call_FA2", ann "migratable"]
+
+type Typename = Text
+type Fieldname = Text
+
+-- Represents a Typescript declaration. For now only contains one that
+-- defines a Type.
+data TsDecl
+  = TsType Text TsTypeDef
+  deriving stock Generic
+
+-- Represents an instance of a Typescript type.
+data TsType
+  = TsNumber
+  | TsString
+  | TsBool
+  | TsUnit
+  | TsOption TsType
+  | TsList TsType
+  | TsMap TsType TsType
+  | TsBigMap TsType TsType
+  | TsCustom Typename
+  deriving stock Generic
+
+-- Implements the conversion of Type to its source code representation.
+instance Buildable TsType where
+  build = \case
+    TsNumber -> "number"
+    TsString -> "string"
+    TsBool -> "boolean"
+    TsUnit -> "void"
```

Comment:
I have mixed feelings about converting a `unit` to `void`. The reason is that we may have `unit` in a container such as `list` or `map` or even on the argument of a `lambda` and it will become an `undefined` in these situations. I think it's better if we had an empty type for this such as `{}`.

Reply:
Have replaced it with '{}'.